File: 2024/src/_14.rs
use std::collections::HashSet;
use crate::utils::get_input_data;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
struct Plot { x: i32, y: i32 }
impl Plot {
    fn new(x: i32, y: i32) -> Self { Self { x, y } }

    fn add_wrapped(&self, other: Plot, bounds: Plot) -> Self {
        Plot {
            x: (self.x + other.x).rem_euclid(bounds.x),
            y: (self.y + other.y).rem_euclid(bounds.y)
        }
    }
}

#[derive(Debug, Clone, Copy)]
struct Robot { pos: Plot, vel: Plot }

fn solve_part_one(data: &[Robot]) -> i32 {
    let bounds = Plot::new(101, 103);
    let cx = bounds.x / 2;
    let cy = bounds.y / 2;
    let mut quadrant_counts = [0; 4];
    for robot in data {
        let final_pos = robot.pos.add_wrapped(Plot::new(100 * robot.vel.x, 100 * robot.vel.y), bounds);
        if final_pos.x == cx || final_pos.y == cy {
            continue;
        } else if final_pos.x < cx && final_pos.y < cy {
            quadrant_counts[0] += 1;
        } else if final_pos.x >= cx && final_pos.y < cy {
            quadrant_counts[1] += 1;
        } else if final_pos.x < cx && final_pos.y >= cy {
            quadrant_counts[2] += 1;
        } else if final_pos.x >= cx && final_pos.y >= cy {
            quadrant_counts[3] += 1;
        }
    }
    quadrant_counts.iter().product()
}

fn solve_part_two(mut data: Vec<Robot>) -> i32 {
    let bounds = Plot::new(101, 103);
    let max_frames = 10000; // Arbitrary number
    let mut compressed = vec![0; max_frames];
    for i in 0..max_frames {
        let x_positions: HashSet<i32> = data.iter().map(|robot| robot.pos.x).collect();
        let y_positions: HashSet<i32> = data.iter().map(|robot| robot.pos.y).collect();
        compressed[i] = x_positions.len() + y_positions.len();
        for robot in &mut data {
            robot.pos = robot.pos.add_wrapped(robot.vel, bounds);
        }
    }
    let (seconds, _) = compressed.iter().enumerate().min_by_key(|&(_, &length)| length).unwrap();
    seconds as i32
}

pub async fn run() -> Result<(), Box<dyn std::error::Error>> {
    let data: Vec<Robot> = get_input_data(14).await?.lines().map(|line| {
        let parts: Vec<&str> = line.split_whitespace().collect();
        let pos: Vec<i32> = parts[0].trim_start_matches("p=").split(',').map(|x| x.parse().unwrap()).collect();
        let vel: Vec<i32> = parts[1].trim_start_matches("v=").split(',').map(|x| x.parse().unwrap()).collect();
        Robot { pos: Plot::new(pos[0], pos[1]), vel: Plot::new(vel[0], vel[1]) }
    }).collect();
    println!("Part One: {}", solve_part_one(&data));
    println!("Part Two: {}", solve_part_two(data));
    Ok(())
}

File: 2018/14.js
import {getInputData} from './utils.js';

const data = Number(await getInputData(14));

function generateRecipes(scoreboard, elf1, elf2) {
	const newRecipeSum = scoreboard[elf1] + scoreboard[elf2];
	const newRecipes = newRecipeSum.toString().split('').map(Number);
	scoreboard.push(...newRecipes);
	elf1 = (elf1 + 1 + scoreboard[elf1]) % scoreboard.length;
	elf2 = (elf2 + 1 + scoreboard[elf2]) % scoreboard.length;
	return {scoreboard, elf1, elf2};
}

function checkTargetSequence(scoreboard, targetSequence) {
	const targetLength = targetSequence.length;
	if (scoreboard.slice(-targetLength).join('') === targetSequence) return scoreboard.length - targetLength;
	if (scoreboard.slice(-(targetLength + 1), -1).join('') === targetSequence) return scoreboard.length - targetLength - 1;
	return null;
}

function solvePartOne() {
	let scoreboard = [3, 7], elf1 = 0, elf2 = 1;
	while (scoreboard.length < data + 10) ({scoreboard, elf1, elf2} = generateRecipes(scoreboard, elf1, elf2));
	return scoreboard.slice(data, data + 10).join('');
}

function solvePartTwo() {
	let scoreboard = [3, 7], elf1 = 0, elf2 = 1;
	const targetSequence = data.toString();
	while (true) {
		({scoreboard, elf1, elf2} = generateRecipes(scoreboard, elf1, elf2));
		const sequenceMatchIndex = checkTargetSequence(scoreboard, targetSequence);
		if (sequenceMatchIndex !== null) return sequenceMatchIndex;
	}
}

console.log(`Part One: ${solvePartOne()}`);
console.log(`Part Two: ${solvePartTwo()}`);